### PR TITLE
Throw error when running assertion on unawaited Promises

### LINKED ISSUE #2207: Add option to throw when passing `expect` a normal Promise
### Are you requesting a feature or reporting a bug?
Feature request, sort of. Also, I’m willing to do the work myself.

### What is the current behavior?

Internally, TestCafe uses a class called [ClientFunctionResultPromise](https://github.com/DevExpress/testcafe/blob/master/src/client-functions/result-promise.js) that extends Promise to allow ClientFunctions to be retried in the [Smart Assertion Query Mechanism](http://devexpress.github.io/testcafe/documentation/test-api/assertions/#smart-assertion-query-mechanism). Most properties on Selector instances return these. It's super useful to pass these to `t.expect` so they get retried until they pass, or a timeout passes, e.g.

```js
const element = Selector('#element');
const thingIsVisible = element.visible; // This is a ClientFunctionResultPromise
await t.expect(thingIsVisible).ok({ timeout: 1000 });
```

However, this can also lead to confusion, because if you pass a normal Promise to `t.expect()`, it will not be awaited or retried; it will use the Promise object itself as the value being asserted. And it’s very easy to fall into this trap, because chaining a `.then` on a ClientFunctionResultPromise returns a normal Promise:

```js
const element = Selector('#element');
const thingIsVisible = element().then(snapshot => snapshot.visible); // This is a Promise
await t.expect(thingIsVisible).ok({ timeout: 1000 });
```

This looks fairly similar, but the assertion **always passes**, even if the thing is never visible, because the Promise object is truthy. So the big problem is that the mistake of passing a normal Promise to `t.expect()` is really harmful with `.ok()`. For any other assertion, you’re likely to catch your mistake, but with `.ok()`, it’s very easy to miswrite an assertion that passes when it shouldn’t.

To make matters worse for those of us who are using TypeScript, the [type definitions](https://github.com/DevExpress/testcafe/blob/master/ts-defs/index.d.ts#L263) for TestCafe don’t expose ClientFunctionResultPromise; instead, it types every ClientFunctionResultPromise as a normal Promise. So these two similar-looking examples above actually look _identical_ to the type system, making this mistake impossible to lint for or guard against with static typings.

### What is the expected behavior?

Two proposals:

1. It seems like passing a normal Promise to `t.expect()` is probably a mistake >99% of the time. Can we somehow add a configuration option that would cause `t.expect(normalPromise)` to throw an error? If that’s too presumptive, what about making `t.expect(normalPromise).ok()` throw an error? It could perhaps be bypassed with something like `.ok({ allowUnawaitedPromises: true })`?

2. If the type definitions exposed `ClientFunctionResultPromise`, we could at least lint for this, or write a type-safe wrapper for `t.expect` or something. Only catch with this is that because TypeScript uses structural typing rather than nominal typing, `ClientFunctionResultPromise` would look identical to `Promise` unless the added methods were made public on the type, and they look like they’re meant to be private. There are some possible hacks around this, but they’d still surface awkwardly in the type system. I think my first proposal is better.

Would you accept a PR for either of these ideas? If the former sounds reasonable, what are your thoughts on an API for it?

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #2207 

I expect you all will have feedback on the message language and the new error class I introduced, `AssertionUnawaitedPromiseError`, as there are still some pending questions from me in #2207.

### CODE REVIEW

**Comment by andrewbranch** [comment]
🤔 ^ this Appveyor failure doesn’t seem related to my PR.

**Comment by AlexanderMoskovkin** [comment]
Hi @andrewbranch,

Thanks a lot for your work! We'll take a look what's wrong there.

File: ts-defs/index.d.ts [hunk 1]
```diff
@@ -1164,6 +1164,11 @@ interface AssertionOptions {
      * selector property or a client function was used in assertion.
      */
     timeout?: number;
+    /**
+     * By default, a Promise is not allowed to be passed to an assertion unless it is a selector property
+     * or the result of a client function. Setting this property to `true` overrides that default.
+     */
+    allowUnawaitedPromise?: true;
```
Comment (AndreyBelym) [inline ts-defs/index.d.ts]: Type must be `boolean`, not `true`.

File: src/api/test-controller/assertion.js [hunk 2]
```diff
@@ -15,6 +15,7 @@ export default class Assertion {
             message = void 0;
         }
 
+        var allowUnawaitedPromise = options.allowUnawaitedPromise;
```
Comment (AndreyBelym) [inline src/api/test-controller/assertion.js]: I think, a blank line after this declaration is required to separate it from multi-line declaration with a ternary operator.

File: src/test-run/commands/options.js [hunk 3]
```diff
@@ -240,13 +240,15 @@ export class AssertionOptions extends Assignable {
         super();
 
         this.timeout = null;
+        this.allowUnawaitedPromise = false;
```
Comment (AndreyBelym) [inline src/test-run/commands/options.js]: Could you please align these lines?
```
        this.timeout               = null;
        this.allowUnawaitedPromise = false;
```

File: src/assertions/executor.js [hunk 4]
```diff
@@ -18,9 +19,14 @@ export default class AssertionExecutor extends EventEmitter {
         this.inRetry   = false;
 
         var fn = getFn(this.command);
-
-        this.fn = this.command.actual instanceof ClientFunctionResultPromise ?
-            this._wrapFunction(fn) : fn;
+        var actualCommand = this.command.actual;
+
+        if (actualCommand instanceof ClientFunctionResultPromise)
+            this.fn =  this._wrapFunction(fn);
```
Comment (AndreyBelym) [inline src/assertions/executor.js]: Remove an extra space here after the assignment operator, please:
```
this.fn = this._wrapFunction(fn)
```

File: src/assertions/executor.js [hunk 5]
```diff
@@ -18,9 +19,14 @@ export default class AssertionExecutor extends EventEmitter {
         this.inRetry   = false;
 
         var fn = getFn(this.command);
-
-        this.fn = this.command.actual instanceof ClientFunctionResultPromise ?
-            this._wrapFunction(fn) : fn;
+        var actualCommand = this.command.actual;
```
Comment (AndreyBelym) [inline src/assertions/executor.js]: Could you please align these lines?
```
        var fn            = getFn(this.command);
        var actualCommand = this.command.actual;
```

File: src/errors/test-run/templates.js [hunk 6]
```diff
@@ -260,5 +260,9 @@ export default {
         Cannot evaluate the "${err.actualValue}" expression in the "${err.argumentName}" parameter because of the following error:
 
         ${err.errMsg}
+    `),
+
+    [TYPE.assertionUnawaitedPromiseError]: err => markup(err, `
+        Attempted to run assertions on a Promise object. Did you forget to await it? If not, pass "{ allowUnawaitedPromise: true }" to the assertion options.
```
Comment (AndreyBelym) [inline src/errors/test-run/templates.js]: @VasilyStrelyaev, please, check the new error message

Comment (VasilyStrelyaev) [reply src/errors/test-run/templates.js]: @AndreyBelym It's okay

**Comment by andrewbranch** [comment]
Thanks for reviewing! Do you need anything else from me?

**Comment by AlexanderMoskovkin** [comment]
@andrewbranch You've done the great job! We will include this to the `0.20.0` version (it's a breaking change because it can break some existing tests. So, we can't add it to `0.19.x`). We may release the `0.19.2` version with bug fixes this week and merge the PR after that. We will be able to create a `dev` release with this changes once we merged it.

**Comment by andrewbranch** [comment]
I see v0.19.2 was released and now this PR has a bunch of conflicts ☹️ 

What's the plan; how can I help?

**Comment by AlexanderMoskovkin** [comment]
The branch is rebased, conflicts are resolved

**Pull request merged**